integrationtest/mapper/pojo-base/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/massindexing/MassIndexingDefaultMonitorIT.java

```diff
@@ -0,0 +1,237 @@
+
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import org.apache.logging.log4j.Level;
+
+class MassIndexingDefaultMonitorIT {
+
+	private static final int COUNT = 100;
+	@RegisterExtension
+	public final BackendMock backendMock = BackendMock.create();
+
+	@RegisterExtension
+	public final StandalonePojoMappingSetupHelper setupHelper =
+			StandalonePojoMappingSetupHelper.withBackendMock( MethodHandles.lookup(), backendMock );
+
+	@RegisterExtension
+	public ExpectedLog4jLog logged = ExpectedLog4jLog.create();
+
+	private final StubLoadingContext loadingContext = new StubLoadingContext();
+
+	private SearchMapping setup(String failureHandler) {
+		backendMock.expectAnySchema( Book.NAME );
+
+		SearchMapping mapping = setupHelper.start()
+				.expectCustomBeans()
+				.withPropertyRadical( EngineSettings.BACKGROUND_FAILURE_HANDLER, failureHandler )
+				.setup( Book.class );
+
+		backendMock.verifyExpectationsMet();
+
+		initData();
+
+		return mapping;
+	}
+
+	@ValueSource(booleans = { true, false })
+	@ParameterizedTest
+	void countOnBeforeType(boolean doCounts) {
+		SearchMapping mapping = setup( null );
+
+		logged.expectEvent( Level.INFO, "Mass indexing complete in ", ". Indexed 100/100 entities" );
+		if ( doCounts ) {
+			logged.expectEvent( Level.INFO, "Mass indexing is going to index 100 entities" ).once();
+		}
+		else {
+			logged.expectEvent( Level.INFO, "Mass indexing is going to index 100 entities" ).never();
+		}
+
+		try ( SearchSession searchSession = mapping.createSession() ) {
+			MassIndexer indexer = searchSession.massIndexer()
+					// Simulate passing information to connect to a DB, ...
+					.context( StubLoadingContext.class, loadingContext );
+
+			CompletableFuture<?> indexingFuture = new CompletableFuture<>();
+			indexingFuture.completeExceptionally( new SimulatedFailure( "Indexing error" ) );
+
+			// add operations on indexes can follow any random order,
+			// since they are executed by different threads
+			BackendMock.DocumentWorkCallListContext expectWorks = backendMock.expectWorks(
+					Book.NAME, DocumentCommitStrategy.NONE, DocumentRefreshStrategy.NONE
+			);
+			for ( int i = 0; i < COUNT; i++ ) {
+				final String id = Integer.toString( i );
+				expectWorks
+						.add( id, b -> b
+								.field( "title", "TITLE_" + id )
+								.field( "author", "AUTHOR_" + id )
+						);
+			}
+
+			// purgeAtStart and mergeSegmentsAfterPurge are enabled by default,
+			// so we expect 1 purge, 1 mergeSegments and 1 flush calls in this order:
+			backendMock.expectIndexScaleWorks( Book.NAME, searchSession.tenantIdentifierValue() )
+					.purge()
+					.mergeSegments()
+					.flush()
+					.refresh();
+
+			try {
+				indexer.monitor( DefaultMassIndexingMonitor.builder().countOnBeforeType( doCounts ).build() )
+						.startAndWait();
```

Comment:
So that works 😄 but ! I'll have to go through the docs + add something to the migration guide and add more tests